File: gamedex-datamodel/src/main/java/com/github/ykrasik/gamedex/datamodel/persistence/Game.java
package com.github.ykrasik.gamedex.datamodel.persistence;

import com.github.ykrasik.gamedex.datamodel.GamePlatform;
import com.github.ykrasik.yava.option.Opt;
import com.gs.collections.api.list.ImmutableList;
import lombok.*;

import java.nio.file.Path;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * @author Yevgeny Krasik
 */
@Value
@Builder
@EqualsAndHashCode(of = "id")
@ToString(of = {"id", "name", "path"})
public class Game implements PathEntity {
    @NonNull private final Id<Game> id;
    @NonNull private final Path path;

    @NonNull private final String metacriticDetailUrl;
    @NonNull private final Opt<String> giantBombDetailUrl;

    @NonNull private final String name;
    @NonNull private final GamePlatform platform;
    @NonNull private final Opt<String> description;
    @NonNull private final Opt<LocalDate> releaseDate;
    @NonNull private final Opt<Double> criticScore;
    @NonNull private final Opt<Double> userScore;

    @NonNull private final LocalDateTime lastModified;

    @NonNull private final ImmutableList<Genre> genres;

    @NonNull private final ImmutableList<Library> libraries;
}

File: gamedex-datamodel/src/main/kotlin/com/github/ykrasik/gamedex/datamodel/persistence/Game.kt
package com.github.ykrasik.gamedex.datamodel.persistence

import com.github.ykrasik.gamedex.common.Id
import org.joda.time.DateTime
import org.joda.time.LocalDate
import java.nio.file.Path

/**
 * User: ykrasik
 * Date: 26/05/2016
 * Time: 15:32
 */
data class Game(
    val id: Id<Game>,
    val path: Path,

    val name: String,
    val description: String?,
    val releaseDate: LocalDate?,

    val criticScore: Double?,
    val userScore: Double?,

    val lastModified: DateTime,

    val metacriticUrl: String,
    val giantBombUrl: String?,

    val genres: List<Genre>,
    val library: Library
) {

    override fun toString() = "Game(id = $id, name = $name, path = $path)"
}